src/PowerShellEditorServices/Language/TokenOperations.cs

```diff
@@ -21,103 +21,140 @@ internal static class TokenOperations
-            foldableRegions.AddRange(
-                MatchCustomCommentRegionTokenElements(tokens, RegionKindRegion)
-            );
-
+            // Given a list of tokens, find the tokens that are comments and
+            // the comment text is either `#region` or `#endregion`, and then use a stack to determine
+            // the ranges they span
+            //
             // Find blocks of line comments # comment1\n# comment2\n...
-            foldableRegions.AddRange(
-                MatchBlockCommentTokenElement(tokens, RegionKindComment)
-            );
-
+            // Finding blocks of comment tokens is more complicated as the newline characters are not
+            // classed as comments.  To workaround this we search for valid block comments (See IsBlockCmment)
+            // and then determine contiguous line numbers from there
+            //
             // Find comments regions <# -> #>
-            foldableRegions.AddRange(
-                MatchTokenElement(tokens, TokenKind.Comment, RegionKindComment)
-            );
+            // Match the token start and end of kind TokenKind.Comment
+            var tokenCommentRegionStack = new Stack<Token>();
+            Token blockStartToken = null;
+            int blockNextLine = -1;
 
-            // Remove any null entries. Nulls appear if the folding reference is invalid
-            // or missing
-            foldableRegions.RemoveAll(item => item == null);
-
-            // Sort the FoldingReferences, starting at the top of the document,
-            // and ensure that, in the case of multiple ranges starting the same line,
-            // that the largest range (i.e. most number of lines spanned) is sorted
-            // first. This is needed to detect duplicate regions. The first in the list
-            // will be used and subsequent duplicates ignored.
-            foldableRegions.Sort();
-
-            // It's possible to have duplicate or overlapping ranges, that is, regions which have the same starting
-            // line number as the previous region. Therefore only emit ranges which have a different starting line
-            // than the previous range.
-            foldableRegions.RemoveAll( (FoldingReference item) => {
-                // Note - I'm not happy with searching here, but as the RemoveAll
-                // doesn't expose the index in the List, we need to calculate it. Fortunately the
-                // list is sorted at this point, so we can use BinarySearch.
-                int index = foldableRegions.BinarySearch(item);
-                if (index == 0) { return false; }
-                return (item.StartLine == foldableRegions[index - 1].StartLine);
-            });
+            for (int index = 0; index < tokens.Length; index++)
+            {
+                Token token = tokens[index];
+                if (token.Kind != TokenKind.Comment) { continue; }
+                // Processing for comment regions <# -> #>
```

Comment:
Insert newline here. It really helps when other people read your code.

Reply:
Fixed